[FLINK-25709] Always delete non-deterministic working directories

tillrohrmann opened the pull request:
This commit introduces the distinction between deterministic and non-deterministic
working directories. The former have an explicit configured ResourceID. The latter
will always be deleted because the assumption is that the user won't want to recover
from them. Otherwise, he would have given a deterministic ResourceID.

zentol commented on flink-tests/src/test/java/org/apache/flink/test/recovery/ClusterEntrypointITCase.java:
```diff
@@ -81,4 +87,45 @@ public void testWorkingDirectoryIsNotDeletedInCaseOfProcessFailure() throws Exce
+        boolean success = false;
+        try {
+            CommonTestUtils.waitUntilCondition(
+                    () -> Files.list(workingDirBase.toPath()).findAny().isPresent(),
```
The stream returned by Files.list needs to be closed.

tillrohrmann replied:
Will do.

zentol commented on flink-tests/src/test/java/org/apache/flink/test/recovery/ClusterEntrypointITCase.java:
```diff
@@ -81,4 +87,45 @@ public void testWorkingDirectoryIsNotDeletedInCaseOfProcessFailure() throws Exce
+
+            final File workingDirectory =
+                    Iterables.getOnlyElement(
+                                    Files.list(workingDirBase.toPath())
```
same as above

tillrohrmann replied:
Will do.

tillrohrmann commented:
Thanks for the review @zentol. I've addressed your comments. I also had to add another fixup commit in order to add the shutdown hook to the `TaskManagerRunner`. The problem was that we initialized several services in the constructor where we also registered the shut down hook. This could then lead to the situation that either the shutdown hook wasn't registered or that the initialization and the shutdown logic was running concurrently. In order to fix this up, I have moved the service initialization out of the constructor into a method that is called by the `start` method.
